dhall-toml/src/Dhall/Toml.hs

```diff
@@ -58,35 +74,92 @@ toToml toml key expr  = case expr of
     Core.NaturalLit a -> return $ insertPrim (Toml.Value.Integer $ toInteger a)
     Core.DoubleLit (DhallDouble a) -> return $ insertPrim (Toml.Value.Double a)
     Core.TextLit (Core.Chunks [] a) -> return $ insertPrim (Toml.Value.Text a)
-    -- TODO: probe the element type, if record then table list else inline list
-    -- Core.ListLit _ a -> Left $ Unimplemented
+    Core.ListLit _ a -> case Seq.lookup 0 a of
+        -- empty array
+        Nothing -> return $ insertPrim (Toml.Value.Array [])
+        -- array of table
+        Just (Core.RecordLit _) -> do
+            tables <- mapM (toToml mempty []) $ toList a
+            let tables' = NonEmpty.fromList tables
+            let key' = Key $ NonEmpty.fromList key
+            return $ Toml.TOML.insertTableArrays key' tables' toml
+        -- inline array
+        Just _ -> do
+            anyList <- mapM toAny $ toList a
+            let arrayEither = Toml.AnyValue.toMArray anyList
+            array <- mapLeft (const $ HeterogeneousArray expr) arrayEither
+            return $ insertPrim array
     Core.RecordLit r ->
         let
-            f toml' (key', val) = toToml toml' [Piece key'] (Core.recordFieldValue val)
+            f curKey toml' (key', val) = toToml toml' (curKey ++ [Piece key']) (Core.recordFieldValue val)
         in
             if null key -- at the top level, we can't have a table
-            then foldM f toml (Map.toList r)
-            else do
-                let (flat, nested) = Map.partition (isFlat . Core.recordFieldValue) r
-                -- the order here is important, at least for testing, because
-                -- the PrefixMap inside TOML is dependent on insert order
-                flatTable   <- foldM f mempty    (Map.toList flat)
-                nestedTable <- foldM f flatTable (Map.toList nested)
-                return $ Toml.TOML.insertTable (Key $ NonEmpty.fromList key) nestedTable toml
+            then foldM (f []) toml (Map.toList r)
+            else let
+                (inline, nested) = Map.partition (isInline . Core.recordFieldValue) r
+            in
+                if null inline
```

Comment:
Similar comment here.  Instead of using `null`, pattern match on `inline` so that you can safely build the `NonEmpty` list later on without using `NonEmpty.fromList`

```haskell
case inline of
    [] -> … {- Handle null case -}
    i : nline -> do
        …
        return (Toml.TOML.insertTable (Key (i :| inline)))
```